File: typescript/food-chain/food-chain.ts
import {isNullOrUndefined} from "util";

class Animal {
    constructor(public name: string, public description: string | undefined = undefined, public suffix: string = "") {}
}

export default class FoodChain {
    private static _animals = [
        new Animal("fly"),
        new Animal("spider", "It wriggled and jiggled and tickled inside her.", " that wriggled and jiggled and tickled inside her"),
        new Animal("bird", "How absurd to swallow a bird!"),
        new Animal("cat", "Imagine that, to swallow a cat!"),
        new Animal("dog", "What a hog, to swallow a dog!"),
        new Animal("goat", "Just opened her throat and swallowed a goat!"),
        new Animal("cow", "I don't know how she swallowed a cow!"),
        new Animal("horse")
    ]

    static verse(v: number): string {
        let foodChain = this._animals.slice(0, v - 1).reverse()
        let lastAnimal = this._animals[v - 1]

        let verse = `I know an old lady who swallowed a ${lastAnimal.name}.\n`
        if (!isNullOrUndefined(lastAnimal.description)) {
            verse += `${lastAnimal.description}\n`
        }

        if (v == FoodChain._animals.length) {
            verse += "She's dead, of course!\n"
        } else {
            for (let animal of foodChain) {
                verse += `She swallowed the ${lastAnimal.name} to catch the ${animal.name}${animal.suffix}.\n`

                lastAnimal = animal
            }
            verse += "I don't know why she swallowed the fly. Perhaps she'll die.\n"
        }
        return verse
    }

    static verses(from: number, to: number): string {
        let song = ""

        for (let v = from; v <= to; v++) {
            song += `${this.verse(v)}\n`
        }

        return song.trim() + "\n"
    }
}

File: scala/food-chain/src/main/scala/FoodChain.scala
object FoodChain {
  private val animals = IndexedSeq(
    Animal("fly", "I don't know why she swallowed the fly. Perhaps she'll die."),
    Animal("spider", "It wriggled and jiggled and tickled inside her.", Some("spider that wriggled and jiggled and tickled inside her")),
    Animal("bird", "How absurd to swallow a bird!"),
    Animal("cat", "Imagine that, to swallow a cat!"),
    Animal("dog", "What a hog, to swallow a dog!"),
    Animal("goat", "Just opened her throat and swallowed a goat!"),
    Animal("cow", "I don't know how she swallowed a cow!"),
    Animal("horse", "She's dead, of course!", terminal = true)
  )
  def recite(from: Int, to: Int): String = {
    val verses = for { v <- from to to } yield verse(v)
    verses.mkString("", "\n", "\n")
  }

  private def verse(v: Int): String = {
    def firstAnimal(animal: Animal) = s"I know an old lady who swallowed a ${animal.name}.\n${animal.comment}\n"
    def rest(prior: Animal, current: Animal) = s"She swallowed the ${prior.name} to catch the ${current.aliasOrName}.\n"
    val chain = animals.take(v).reverse
    if (chain.head.terminal || chain.length == 1) firstAnimal(chain.head)
    else {
      val (verses, _) = chain.foldLeft(("", chain.head)) { case ((verse, prior), current) =>
        if (current == prior) (firstAnimal(current), current)
        else (verse + rest(prior, current), current)
      }

      verses + animals.head.comment + "\n"
    }
  }
}

case class Animal(name: String, comment: String, alias: Option[String] = None, terminal: Boolean = false) {
  def aliasOrName = alias.getOrElse(name)
}